### PR TITLE
Cmake arch fix

### LINKED ISSUE #2384: CMake Helper Enhancement - Implicit Addition of Known CLI Build Flags
A number of users and recipe's have implemented code such as that shown below, in order to pass build flags into CMake via the CLI when using Conan's CMake helper.  After discussion with Conan team, we've established that the flags are already passed in via the `conanbuildinfo.cmake` file and `conan_basic_setup()` steps, which seem to make these build flags redundant.  However, as indicated by @SSE4, some flags need to be passed to CMake via CLI so that they are received and processed by CMake at an earlier stage.  

Thus, this is a feature request to add implicit passing of certain flags by Conan's CMake helper, to CMake, via the CLI.  The example shows the `CMAKE_C_FLAGS` and `CMAKE_CXX_FLAGS` flags, but we should probably check CMake documentation/experts to see if there are any other nuances with other flags. 

Here is the original explanation: 
https://github.com/bincrafters/community/issues/82#issuecomment-359656248

Here is an example of the kind of code we want to eliminate from recipe's:   
```python
        if self.settings.os == 'Linux':
            if self.settings.arch == 'x86':
                cmake.definitions['CMAKE_C_FLAGS'] = '-m32'
                cmake.definitions['CMAKE_CXX_FLAGS'] = '-m32'
            elif self.settings.arch == 'x86_64':
                cmake.definitions['CMAKE_C_FLAGS'] = '-m64'
                cmake.definitions['CMAKE_CXX_FLAGS'] = '-m64'
```

Here is the recipe this came from: 
https://github.com/bincrafters/conan-sdl2/blob/testing/2.0.7/conanfile.py#L158

- [X] I've read the [CONTRIBUTING guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [X] I've specified the Conan version, operating system version and any tool that can be relevant.
- [ X I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

### PR DESCRIPTION
**Pull request by SSE4**
- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [x] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

Closes #2384 

adds centralized helpers to define compiler/architecture specific flags (kind of refactoring).
advantages:
* less code duplication across the conan codebase
* allow recipe and generator authors to re-use common code to define flags
* abstract/hide unnecessary details

### CODE REVIEW

File: conans/client/tools/compiler_flags.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
```
Comment (memsharded) [inline conans/client/tools/compiler_flags.py]: Is this mostly a tools for users in their conanfiles, or for use by build helpers? I'd say the second, so probably should go to ``conans/build`` instead.

File: conans/client/build/cmake.py [hunk 2]
```diff
@@ -247,6 +250,18 @@ def build_config(self):
         return ""
 
     def _get_cmake_definitions(self):
+        def add_cmake_flags(cmake_flags, name, flags):
+            """
+            appends compiler linker flags (if already present), or just sets
+            """
+            if len(flags):
+                joint_flags = ' '.join(flags)
+                if name not in cmake_flags.keys():
```
Comment (memsharded) [inline conans/client/build/cmake.py]: ``keys()`` redundant

File: conans/client/build/cmake.py [hunk 3]
```diff
@@ -247,6 +250,18 @@ def build_config(self):
         return ""
 
     def _get_cmake_definitions(self):
+        def add_cmake_flags(cmake_flags, name, flags):
+            """
+            appends compiler linker flags (if already present), or just sets
+            """
+            if len(flags):
```
Comment (memsharded) [inline conans/client/build/cmake.py]: ``if flags:``

Comment (SSE4) [reply conans/client/build/cmake.py]: done

File: conans/client/build/cmake.py (hunk 2 above)
Comment (SSE4) [reply conans/client/build/cmake.py]: done

File: conans/client/tools/compiler_flags.py (hunk 1 above)
Comment (SSE4) [reply conans/client/tools/compiler_flags.py]: moved to conans.client.build

File: conans/client/tools/compiler_flags.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+# -*- coding: utf-8 -*-
+
+
+class CompilerFlags:
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/tools/compiler_flags.py]: done

File: conans/client/tools/compiler_flags.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        self.cxxflags.extend(additional_flags.cxxflags)
+        self.ldflags.extend(additional_flags.ldflags)
+        self.defines.extend(additional_flags.defines)
+
+
+def architecture_flags(compiler=None, arch=None):
+    """
+    returns flags specific to the target architecture and compiler
+    """
+    if str(compiler) in ['gcc', 'clang', 'sun-cc']:
```
[the remark this answers is not in the document]
Comment (SSE4) [reply conans/client/tools/compiler_flags.py]: added apple-clang. note that AutoToolsBuildEnvironment used to apply flags for clang and apple-clang, while CMake helper used to apply only for clang (or more precisely, it used os check, and didn't include MacOSX). now behavior is consistent.

**Comment by lasote** [comment]
we have some conflicts for other merged PR, could you take a look?

**Comment by SSE4** [comment]
@lasote  okay

**Comment by SSE4** [comment]
@lasote done

File: conans/client/build/autotools_environment.py [hunk 6]
```diff
@@ -4,39 +4,17 @@
...
 
-sun_cc_libcxx_flags_dict = {"libCstd": "-library=Cstd",
-                            "libstdcxx": "-library=stdcxx4",
-                            "libstlport": "-library=stlport4",
-                            "libstdc++": "-library=stdcpp"}
-
-architecture_dict = {"x86_64": "-m64", "x86": "-m32"}
-
 
 def stdlib_flags(compiler, libcxx):
```
Comment (lasote) [inline conans/client/build/autotools_environment.py]: I think we could remove the `stdlib_flags` and `stdlib_defines` functions and call directly to the libcxx_flags, right?

Comment (SSE4) [reply conans/client/build/autotools_environment.py]: yes, if it makes sense

File: conans/client/tools/compiler_flags.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (memsharded) [reply conans/client/tools/compiler_flags.py]: Perfect, thanks for addressing all these things so quickly!

File: conans/client/build/autotools_environment.py (hunk 6 above)
Comment (SSE4) [reply conans/client/build/autotools_environment.py]: removed

**Comment by SSE4** [comment]
I am going to merge it with #2325 extracting methods to format defines, include and library paths, as well as adding other flags (debug, sysroot, etc)

**Comment by SSE4** [comment]
@memsharded @lasote I've rebased on latest develop, merged changes from #2325 here and modeled most of the compiler-specific settings into the new helpers, please take a look

**Comment by memsharded** [comment]
It got some conflicts with your other PR just merged. I have had a look to check if I could quickly resolve the conflicts, but seems it requires careful checking.

**Comment by SSE4** [comment]
@memsharded I'll take a look, thanks

**Comment by SSE4** [comment]
@memsharded @lasote  rebased one more time, pls review

File: conans/client/build/compiler_flags.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+    return ["-D%s" % define for define in defines]
+
+
+def format_include_paths(include_paths, win_bash=False, subsystem=None, compiler=None):
+    return ["-I%s" % adjust_path(include_path, win_bash=win_bash, subsystem=subsystem, compiler=compiler)
+            for include_path in include_paths]
+
+
+def format_library_paths(library_paths, win_bash=False, subsystem=None, compiler=None):
+    pattern = "-LIBPATH:%s" if str(compiler) == 'Visual Studio' else "-L%s"
```
Comment (lasote) [inline conans/client/build/compiler_flags.py]: Why -LIBPATH and not /LIBPATH? is a bug or there is a reason?

File: conans/client/build/compiler_flags.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        return CompilerFlags(cflags=sysroot_flag, cxxflags=sysroot_flag, ldflags=sysroot_flag)
+    return CompilerFlags()
+
+
+def format_defines(defines):
+    return ["-D%s" % define for define in defines]
+
+
+def format_include_paths(include_paths, win_bash=False, subsystem=None, compiler=None):
+    return ["-I%s" % adjust_path(include_path, win_bash=win_bash, subsystem=subsystem, compiler=compiler)
```
Comment (lasote) [inline conans/client/build/compiler_flags.py]: Why -I and not /I with visual?

Comment (SSE4) [reply conans/client/build/compiler_flags.py]: Visual Studio accepts both forms -l and /l, -LIBPATH and /LIBPATH, it doesn't really matter a lot., we can use any of two options. change or keep?

Comment (lasote) [reply conans/client/build/compiler_flags.py]: Keep it. Np. I will open a PR for you the later on Monday. Thanks for your quick answer

**Comment by lasote** [comment]
Hi @SSE4 I've opened a new Pr containing all your commits (won't be squashed), please take a look and tell me if you see something. #2458

**Pull request merged**